cdap-authorization-ldap-role/src/main/java/io/cdap/cdap/security/authorization/ldap/role/searcher/LDAPSearcher.java

```diff
@@ -0,0 +1,147 @@
+import java.util.Hashtable;
+import java.util.Set;
+import java.util.stream.Collectors;
+import javax.naming.Context;
+import javax.naming.NamingEnumeration;
+import javax.naming.NamingException;
+import javax.naming.PartialResultException;
+import javax.naming.directory.Attribute;
+import javax.naming.directory.DirContext;
+import javax.naming.directory.SearchControls;
+import javax.naming.directory.SearchResult;
+
+import static javax.naming.directory.SearchControls.SUBTREE_SCOPE;
+
+
+/**
+ * Service for searching user's groups in LDAP
+ */
+public class LDAPSearcher {
+  private static final Logger LOG = LoggerFactory.getLogger(LDAPSearcher.class);
+
+  private static final String LDAP_SOCKET_FACTORY = "java.naming.ldap.factory.socket";
+  private static final String LDAPS_PROTOCOl = "ldaps";
+  private static final String BASE_DN_SPLITTER = ";";
+  private static final int MAX_RETRIES = 5;
+
+  private final LDAPSearchConfig config;
+  private final String[] baseDNList;
+
+  public LDAPSearcher(LDAPSearchConfig config) {
+    this.config = config;
+    baseDNList = config.getSearchBaseDn().split(BASE_DN_SPLITTER);
+  }
+
+  public void testConnection() {
+    try {
+      DirContext context = getConnection();
+      context.close();
+    } catch (NamingException e) {
+      LOG.error("Failed to establish connection to '{}'", config.getUrl(), e);
+      throw new RuntimeException(e);
+    }
+  }
+
+  public Set<String> searchGroups(String username) {
+    Set<String> groups = Collections.emptySet();
+
+    try {
+      DirContext context = getConnection();
+
+      SearchControls controls = new SearchControls();
+      if (config.isRecursiveSearch()) {
+        controls.setSearchScope(SUBTREE_SCOPE);
+      }
+
+      groups = Arrays.stream(baseDNList)
+        .map(baseDN -> searchGroups(baseDN, username, context, controls))
+        .flatMap(Collection::stream)
+        .collect(Collectors.toSet());
+
+      context.close();
+    } catch (NamingException e) {
+      LOG.error("Failed to find groups for '{}'", username, e);
+    }
+
+    return groups;
+  }
+
+  private Set<String> searchGroups(String baseDN, String username, DirContext context, SearchControls controls) {
+    String filter = String.format(config.getSearchFilter(), username);
+    Set<String> groups = new HashSet<>();
+
+    try {
+      NamingEnumeration<SearchResult> renum = context.search(baseDN, filter, controls);
+
+      if (!renum.hasMore()) {
+        LOG.warn("Cannot locate user information for '{}'", username);
+        return groups;
+      }
+
+      SearchResult result = renum.next();
+
+      Attribute memberOf = result.getAttributes().get(config.getMemberAttribute());
+      if (memberOf != null) {
+        for (int i = 0; i < memberOf.size(); i++) {
+          groups.add(memberOf.get(i).toString());
+        }
+      }
+    } catch (PartialResultException e) {
+      LOG.warn("Failed to find groups for '{}' in '{}'", username, baseDN);
```

Comment:
definitely should not just ignore an exception

Reply:
it is case when there is nor groups in current search dn, they can be in another as we support multiple

Reply:
in this case it should not be a WARN. May be a DEBUG

Reply:
done

Reply:
done